core/trino-main/src/test/java/io/trino/execution/TestSqlTaskManagerRaceWithCatalogPrune.java

```diff
@@ -0,0 +1,255 @@
+        }
+    };
+    private final ThreadPoolExecutor threadPoolExecutor = new ThreadPoolExecutor(2, 2, 10, TimeUnit.MINUTES, new LinkedBlockingDeque<>());
+    private final AtomicInteger sequence = new AtomicInteger(1);
+
+    @Test
+    public void testMultipleTaskUpdatesWithMultipleCatalogPrunes()
+    {
+        ConnectorServicesProvider workerConnectorServiceProvider = new WorkerDynamicCatalogManager(MOCK_CATALOG_FACTORY);
+        SqlTaskManager workerTaskManager = getWorkerTaskManagerWithConnectorServiceProvider(workerConnectorServiceProvider);
+
+        CatalogPruneTask catalogPruneTask = new TestingLocalCatalogPruneTask(
+                new NoInfoTransactionManager(),
+                NO_CATALOGS,
+                NOOP_CONNECTOR_SERVICES_PROVIDER,
+                new NodeInfo("testversion"),
+                new CatalogPruneTaskConfig(),
+                workerTaskManager);
+
+        Future<Void> catalogTaskFuture = Futures.submit(() ->
+        {
+            for (int i = 0; i < NUM_TASKS; i++) {
+                String catalogName = "catalog_" + randomNameSuffix();
+                CatalogHandle catalogHandle = createRootCatalogHandle(catalogName, new CatalogHandle.CatalogVersion(UUID.randomUUID().toString()));
+                workerTaskManager.updateTask(
+                        TestingSession.testSession(),
+                        newTaskId(),
+                        Span.getInvalid(),
+                        Optional.of(fragmentWithCatalog(catalogHandle)),
+                        ImmutableList.of(new SplitAssignment(TABLE_SCAN_NODE_ID, ImmutableSet.of(), true)),
+                        PipelinedOutputBuffers.createInitial(PARTITIONED).withBuffer(OUT, 0).withNoMoreBufferIds(),
+                        ImmutableMap.of(),
+                        false);
+                try {
+                    Thread.sleep(0, ThreadLocalRandom.current().nextInt(25, 75));
+                }
+                catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+                assertDoesNotThrow(() -> workerConnectorServiceProvider.getConnectorServices(catalogHandle));
+            }
+        }, threadPoolExecutor);
+
+        Future<Void> pruneCatalogsFuture = Futures.submit(() ->
+        {
+            for (int i = 0; i < NUM_TASKS; i++) {
+                catalogPruneTask.pruneWorkerCatalogs();
+                try {
+                    Thread.sleep(0, ThreadLocalRandom.current().nextInt(25, 75));
+                }
+                catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+            }
+        }, threadPoolExecutor);
+
+        assertDoesNotThrow(() -> catalogTaskFuture.get(120, TimeUnit.SECONDS));
+        assertDoesNotThrow(() -> pruneCatalogsFuture.get(120, TimeUnit.SECONDS));
+    }
+
+    @AfterAll
+    public void cleanup()
```

Comment:
Place this above the test... normally these go after the constructor or setup method and before the tests